drivers/tsl4531x/tsl4531x.c

```diff
@@ -0,0 +1,197 @@
+                       0) < 0)) {
+
+        DEBUG("[Error] Cannot configure device.\n");
+        i2c_release(params->i2c_dev);
+        return -ENXIO;
+    }
+
+    /* If device was configured correctly, initialise the device descriptor */
+    dev->i2c_dev = params->i2c_dev;
+    dev->i2c_addr = params->i2c_addr;
+    dev->integration_time = params->integration_time;
+    dev->low_power_mode = params->low_power_mode;
+    dev->sample_start_time = 0; /* Device assumed to start up at same time as
+                                   microcontroller - i.e. when it hits this line */
+
+    i2c_release(params->i2c_dev);
+
+    return 0;
+}
+
+int tsl4531x_set_low_power_mode(tsl4531x_t *dev, bool low_power_mode)
+{
+    assert(dev);
+
+    dev->low_power_mode = low_power_mode;
+
+    i2c_acquire(dev->i2c_dev);
+    i2c_write_reg(dev->i2c_dev,
+                  dev->i2c_addr,
+                  TSL4531X_COMMAND(TSL4531X_CONTROL_REG),
+                  TSL4531X_CONTROL(TSL4531X_POWER_MODE(low_power_mode)),
+                  0);
+    i2c_release(dev->i2c_dev);
+
+    /* In high power mode only, we restart the sample ready timer, because only
+       in this mode it's used to indicate readiness after startup. */
+    if (!dev->low_power_mode) {
+        dev->sample_start_time = xtimer_now_usec();
+    }
+
+    return 0;
+}
+
+int tsl4531x_start_sample(tsl4531x_t *dev)
+{
+    assert(dev);
+
+    /* Don't change the mode to one-shot if the device is in high power mode. */
+    if (dev->low_power_mode) {
+        _start_sample(dev);
+        dev->sample_start_time = xtimer_now_usec();
+    }
+
+    return 0;
+}
+
+uint32_t tsl4531x_time_until_sample_ready(const tsl4531x_t *dev)
+{
+    assert(dev);
+
+    uint32_t elapsed = (xtimer_now_usec() - dev->sample_start_time);
+
+    return (TSL4531X_GET_INTEGRATION_TIME_USEC(dev->integration_time, TSL4531X_PSAVESKIP_ON) > elapsed ?
+            TSL4531X_GET_INTEGRATION_TIME_USEC(dev->integration_time, TSL4531X_PSAVESKIP_ON) - elapsed :
+            0);
+}
+
+int tsl4531x_get_sample(const tsl4531x_t *dev)
+{
+    assert(dev);
+
+    uint8_t als_data[2]; /* = {[DATALOW], [DATAHIGH]} */
+
+    i2c_acquire(dev->i2c_dev);
+    i2c_read_regs(dev->i2c_dev,
+                  dev->i2c_addr,
+                  TSL4531X_COMMAND(TSL4531X_ALSDATA1_REG),
+                  als_data, 2, 0);
+    i2c_release(dev->i2c_dev);
+
+    return MULTIPLY_DATA((((uint16_t)als_data[_DATAHIGH]) << 8)
+                         + als_data[_DATALOW], dev->integration_time);
+}
+
+int tsl4531x_simple_read(const tsl4531x_t *dev)
+{
+    assert(dev);
+
+    uint32_t t;
+
+    if (dev->low_power_mode) {
+        /* Call the const *dev version, so this basic read also takes a const *dev
+           and therefore is compatible with the SAUL API. */
+        t = _start_sample(dev);
+    }
+    else {
+        /* Wait until the device has collected at least one sample. Note that
```

Comment:
This is not nice. There should be a way to fix this.

Reply:
As discussed offline, I added a flag to indicate whether the device has started up. this is set if
a) the sample ready countdown timer has hit zero, and
b) the sample has been checked (necessary to detect the device having been started up)

The read is now non-const, and in the saul read function I've handed a pointer to a dev pointer which gets around saul's requirement for a const pointer. (i.e. the pointer pointer is const, but the dev pointer isn't)